packages/ddp-client/livedata_connection.js

```diff
@@ -1206,13 +1219,52 @@ _.extend(Connection.prototype, {
       // We'll now process and all of our buffered messages, reset all stores,
       // and apply them all at once.
       _.each(self._messagesBufferedUntilQuiescence, function (bufferedMsg) {
-        self._processOneDataMessage(bufferedMsg, updates);
+        self._processOneDataMessage(bufferedMsg, self._bufferedWrites);
       });
       self._messagesBufferedUntilQuiescence = [];
     } else {
-      self._processOneDataMessage(msg, updates);
+      self._processOneDataMessage(msg, self._bufferedWrites);
+    }
+
+    // Immediately flush writes when:
+    //  1. Buffering is disabled. Or;
+    //  2. any non-(added/changed/removed) message arrives.
+    var standardWrite = _.include(['added', 'changed', 'removed'], msg.msg);
+    if (self._bufferedWritesInterval === 0 || !standardWrite) {
+      self._flushBufferedWrites();
+      return;
+    }
+
+    if (self._bufferedWritesFlushAt === null) {
+      self._bufferedWritesFlushAt = new Date().valueOf() + self._bufferedWritesMaxAge;
+    }
+    else if (self._bufferedWritesFlushAt < new Date().valueOf()) {
+      self._flushBufferedWrites();
+      return;
+    }
+
+    if (self._bufferedWritesFlushHandle) {
+      clearTimeout(self._bufferedWritesFlushHandle);
+    }
+    self._bufferedWritesFlushHandle = setTimeout(self.__flushBufferedWrites,
+                                                      self._bufferedWritesInterval);
+  },
+
+  _flushBufferedWrites: function () {
+    var self = this;
+    if (self._bufferedWritesFlushHandle) {
+      clearTimeout(self._bufferedWritesFlushHandle);
+      self._bufferedWritesFlushHandle = null;
     }
 
+    self._bufferedWritesFlushAt = null;
+    self._performWrites(self._bufferedWrites);
```

Comment:
OK, I've briefly looked into the failing test.

These two lines are safe in normal JS but not in fiberized JS.  Specifically, it's possible for more messages to come down the network before `_performWrites` returns, which can lead to sending the same buffered writes to `_performWrites` again!

This is what is happening in the failing test.  Specifically, the structure is such that for some reason synchronously to `_performWrites` we call the `onReady` callback for a subscription (this is via `_runAfterUpdateCallbacks`).  In the case of the failing test, that onReady is a testAsyncMulti `expect` function, which believe it or not actually ends up invoking the next step of the test synchronously.  This means that `_performWrites` never returns in this case.

Changing these two lines to
```
    var writes = self._bufferedWrites;
    self._bufferedWrites = {};
    self._performWrites(writes);
```
fixes the failing test.  It does leave one wondering if there is a further problem.

I think in the old code it was moderately unproblematic that `_runAfterUpdateCallbacks` could block, since it was the final thing called from the "receive message" callback.  But now it can also be called from near the beginning of `_livedata_result`, which is a little hairier.  It's a little bad that user code (sub ready callbacks and method call callbacks) which can block will prevent return values from being received (including the onResultReceived callback which is explicitly supposed to be the one that doesn't have to wait on other stuff).  But I guess that's relatively unavoidable when you're using Fibers.

I would definitely make the change I suggested above.  I'm not sure if there are deeper changes needed to make this stuff more bulletproof — but maybe people don't use minimongo connected to DDP clients too much in Node anyway.